Bluetooth: Host: Add LE Power Control Request Procedure APIs

KyraLengfeld opened the pull request:
This commit adds the LE API's for the LE Power Control Request Feature in Zephyr.

The support of feature is provided with the controller-based feature selection with BT_CTLR_LE_POWER_CONTROL_SUPPORT and is selectable via BT_TRANSMIT_POWER_CONTROL.

With the new APIs, the applications will:
get improved reading of local and remote tx power
be aware of changes in remote or local tx power

Defined HCI commands in Core Spec v5.4:
7.8.117 LE Enhanced Read Transmit Power Level command: improvement to existing local tx power reading.
7.8.118 LE Read Remote Transmit Power Level command: Remote tx power is read through an event (LE Transmit Power Reporting) 7.8.121 LE Set Transmit Power Reporting Enable command: Enables local or remote tx power reporting to monitor changes in tx power 7.7.65.33 LE Transmit Power Reporting event

Note: to utilize the Feature fully, Nordic-LL-only vendor-specific commands are needed. These will not be added in RTOS zephyr but instead implemented in a maintainable way in sdk.

KyraLengfeld committed:
Bluetooth: Host: Add LE Power Control Request Procedure APIs

This commits adds the LE API's for the LE Power Control Request Feature
in Zephyr.

The support of feature is provided with the controller-based feature
selection with BT_CTLR_LE_POWER_CONTROL_SUPPORT and is selectable via
BT_TRANSMIT_POWER_CONTROL.

With the new APIs, the applications will:
get improved reading of local and remote tx power
be aware of changes in remote or local tx power

Defined HCI commands in Core Spec v5.4:
7.8.117 LE Enhanced Read Transmit Power Level command:
improvement to existing local tx power reading.
7.8.118 LE Read Remote Transmit Power Level command:
Remote tx power is read through an event (LE Transmit Power Reporting)
7.8.121 LE Set Transmit Power Reporting Enable command:
Enables local or remote tx power reporting to monitor changes in tx power
7.7.65.33 LE Transmit Power Reporting event

Note: to utilize the Feature fully Nordic-LL-only vendor-specific commands
are needed. These will not be added in RTOS zephyr but instead
implemented in a maintainable way in sdk.

Signed-off-by: Kyra Lengfeld <kyra.lengfeld@nordicsemi.no>

alwa-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1052,6 +1131,22 @@ struct bt_conn_cb {
+	 *  @param report Transmit power report.
+	 */
+	void (*tx_power_report)(struct bt_conn *conn,
+				      struct bt_conn_le_tx_power_report *report);
```
Since there the same object is passed to multiple listeners, we should enforce constness.

```suggestion
				      const struct bt_conn_le_tx_power_report *report);
```

alwa-nordic commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -484,6 +487,14 @@ config BT_CTLR_CONN_RSSI
+config BT_CTLR_LE_POWER_CONTROL
+	bool "LE Power Control Request Feature"
+	depends on BT_CTLR_LE_POWER_CONTROL_SUPPORT
+	default y if BT_TRANSMIT_POWER_CONTROL
```
There is both a `default` here and a `select` in `BT_TRANSMIT_POWER_CONTROL`. Is it correct to have both?

KyraLengfeld replied:
You are right, best case overkill, worst case source of kconfig issues in future. I will remove the "default y if".

wiba-nordic commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -750,6 +750,14 @@ config BT_CONN_PARAM_RETRY_TIMEOUT
+	select BT_CTLR_LE_POWER_CONTROL
+	help
+	  Enable support for controlling transmit power. The controller shall
+	  support LE Power Control Request feature that defined in the Bluetooth
```
```suggestion
	  support LE Power Control Request feature that is defined in the Bluetooth
```

jhedberg commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1052,6 +1131,22 @@ struct bt_conn_cb {
+	 *  @param conn Connection object.
+	 *  @param report Transmit power report.
+	 */
+	void (*tx_power_report)(struct bt_conn *conn,
+				      const struct bt_conn_le_tx_power_report *report);
```
minor: the split line doesn't seem to be aligned consistently (to the other similar places in this file, or Bluetooth subsystem code for that matter). Should be n tabs + 0-7 spaces to align right after the `(` of the previous line.

jhedberg replied:
FWIW, this seems to be the only place with this issue - the alignment looks correct in all other "split line" places in this PR

KyraLengfeld replied:
Will change soon. Thanks :)

jhedberg commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -479,6 +479,50 @@ struct bt_conn_le_tx_power {
+	/** Remote Transmit power changed. */
+	BT_CONN_LE_TX_POWER_REPORT_REASON_REMOTE_CHANGED = 1,
+
+	/** HCI_LE_Read_Remote_Transmit_Power_Level command completed. */
```
The public `conn.h` API isn't really tied to HCI in any way (although it might reuse defined values like I mentioned in my other comment), so I'm not sure a reference to a HCI command is appropriate.

KyraLengfeld replied:
I can reference `bt_conn_le_get_remote_tx_power_level` instead? Otherwise I can describe it.

jhedberg replied:
> I can reference `bt_conn_le_get_remote_tx_power_level` instead? Otherwise I can describe it.

Either way is fine, thanks.

KyraLengfeld replied:
I left that comment as is, as I shifted it into hci_types.h (Which I assume you meant in your other comment.).

KyraLengfeld commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -479,6 +479,50 @@ struct bt_conn_le_tx_power {
 	int8_t max_level;
 };
 
+enum bt_conn_le_tx_power_report_reason {
+	/** Local Transmit power changed. */
+	BT_CONN_LE_TX_POWER_REPORT_REASON_LOCAL_CHANGED = 0,
+
+	/** Remote Transmit power changed. */
+	BT_CONN_LE_TX_POWER_REPORT_REASON_REMOTE_CHANGED = 1,
+
+	/** HCI_LE_Read_Remote_Transmit_Power_Level command completed. */
+	BT_CONN_LE_TX_POWER_REPORT_REASON_READ_REMOTE_COMPLETED = 2,
+};
```
I feel it is beneficial to have a description of the values defined like this, so I would just add a comment referencing the HCI specification? Thinking about your other comment this would be a bit contradictory to reference though(as conn.h is not tied to hci).

jhedberg replied:
> I feel it is beneficial to have a description of the values defined like this, so I would just add a comment referencing the HCI specification? Thinking about your other comment this would be a bit contradictory to reference though(as conn.h is not tied to hci).

I agree, it's beneficial. There should at least be a comment for the `reason` struct member that it contains HCI values. As for the values themselves, I'm thinking maybe `hci.h` is a better place for them (near the command response parameter definition) rather than `conn.h`?

KyraLengfeld replied:
I added a core spec reference, imo core reference is the cleanest way to show where it comes from and the user can easily determine if it is outdated or not. If you though about something else, let me know :)

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -571,6 +571,46 @@ struct bt_hci_rp_read_tx_power_level {
+#define BT_HCI_LE_TX_POWER_REPORT_ENABLE        0x01
+#define BT_HCI_OP_LE_SET_TX_POWER_REPORT_ENABLE BT_OP(BT_OGF_LE, 0x007A)
+
+enum bt_conn_le_tx_power_report_reason {
+	/** Local Transmit power changed. */
+	BT_CONN_LE_TX_POWER_REPORT_REASON_LOCAL_CHANGED = 0,
+
+	/** Remote Transmit power changed. */
+	BT_CONN_LE_TX_POWER_REPORT_REASON_REMOTE_CHANGED = 1,
+
+	/** HCI_LE_Read_Remote_Transmit_Power_Level command completed. */
+	BT_CONN_LE_TX_POWER_REPORT_REASON_READ_REMOTE_COMPLETED = 2,
+};
```
I guess this should be moved closer to `struct bt_hci_evt_le_transmit_power_report` since that's where you have `uint8_t reason`? Also, since it's `uint8_t reason` and not `enum bt_conn_le_tx_power_report_reason` I'd stick to defines, like everything else in this header file, and the prefix would then be `BT_HCI_*` instead of `BT_CONN_*` or `bt_conn_*` (if you check, there aren't any other definitions in this header file that don't use the bt_hci prefix, so already from that perspective it's ill-suited here)

KyraLengfeld replied:
I see your points. Reason why I placed it there was as we technically only use it in conn.h and in connection with read remote tx power. 
How about changing the report type to `enum bt_conn_le_tx_power_report_reason` (I would rename it to `enum bt_hci_le_tx_power_report_reason` ), actually using that in `bt_hci_evt_le_transmit_power_report` (as then it is defined right above, would be kind-off ugly imo to not utilize it), and keep utilizing it in `bt_conn_le_tx_power_report` in conn.h? Yes we would have an hci type (`enum bt_hci_le_tx_power_report_reason`) in conn.h, but we reference the hci spec regarding reporting reason above that parameter anyways.

KyraLengfeld replied:
See the new changes for how I meant this.

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -2905,6 +2935,28 @@ struct bt_hci_evt_le_req_peer_sca_complete {
+	/** HCI_LE_Read_Remote_Transmit_Power_Level command completed. */
+	BT_HCI_LE_TX_POWER_REPORT_REASON_READ_REMOTE_COMPLETED = 2,
+};
+struct bt_hci_evt_le_transmit_power_report {
+	uint8_t	 status;
+	uint16_t handle;
+	enum bt_hci_le_tx_power_report_reason  reason;
```
I don't think this is safe, and may not always work. Most platforms will use the native word width (e.g. 32 bits) for enum types. The only way to force the compiler to shrink the type to the minimum possible width is with the `__packed` annotation. Since these HCI structs themselves have a `__packed` annotation it *might* get "inherited" to the individual members, but I suspect it might also just be used to strip off unnecessary padding. Either way, this is not ideal: for these protocol structs we really do want to be explicit about the width of each member, so that the exactly match what's defined in the HCI specification, i.e. `uint8_t`, `uint16_t`, etc are really the appropriate types IMO.

KyraLengfeld replied:
Hm, yeah, I see. But then what do you think I should do about the enum? have it be a bt_conn enum, with BT_HCI types, that is defined in hci_types.h but only used in conn.h?

jhedberg replied:
I agree that enums are nice when they can be used, however in this case, if you want to be consistent between the HCI struct and the API, both would need to use `uint8_t`, and in that case you're back to my proposal of using `#define` :)

KyraLengfeld replied:
Alright :) In that case there will be no usage of these defines implemented though, don't see the benefit to have them defined then. Instead I'd just reference the spec and perhaps expand the explanation in the comment before the parameter. 
Let me know if I am missing something in that thought process.

jhedberg replied:
> Alright :) In that case there will be no usage of these defines implemented though, don't see the benefit to have them defined then. Instead I'd just reference the spec and perhaps expand the explanation in the comment before the parameter. Let me know if I am missing something in that thought process.

That's fine, though I do think they might be helpful to applications, so they don't need to hard-code values or create their own defines. In fact, this whole new feature submission should probably have been accompanied with a sample app that shows the intended usage of the new API. I.e. it's not just that there's no usage of the defines as the PR now stands, there's no usage of any part of the API, which is not really good when you consider of CI coverage and sample application code.

KyraLengfeld replied:
I agree with that, but the feature can only be utilized minimally with the le commands, the real benefit of it comes with the nordic-only vendor-specific commands.

jhedberg replied:
> I agree with that, but the feature can only be utilized minimally with the le commands, the real benefit of it comes with the nordic-only vendor-specific commands.

If there's no value to non-Nordic platforms, then the feature doesn't belong upstream. If there is some value, sure, but you need to demonstrate that the code actually works. Is building the new code currently covered by any upstream test cases? In other words, when I see the CI checks passing has it actually tried to build the new code? We also have the bsim tests which could be used to implement runtime CI checks for the feature and have CI verify that the code doesn't just build but actually works.

jhedberg commented:
Another thing I realized: it would be good to split the controller changes into a separate commit from the host ones. And if you add a sample application + CI integration (which I think you should) that should also be in a separate (third) commit.

KyraLengfeld commented:
> Another thing I realized: it would be good to split the controller changes into a separate commit from the host ones. And if you add a sample application + CI integration (which I think you should) that should also be in a separate (third) commit.

This is the LE part of changes, which we want eventually in the fork sdk-zephyr, but since the API's are basically just calling hci commands as defined by spec, it is not something sdk-specific and would expand rtos-zephyr's capabilities.
(See [Draft PR of sdk-nrf](https://github.com/nrfconnect/sdk-nrf/pull/13136) changes, the description/commit message should give a good summary of it.)

I did manual testing with bt shell (Which I can provide to you, or add as a commit for le commands here. Don't think that is needed to merge to main though).
There is also the integration of the enhanced get tx power command [here](https://github.com/zephyrproject-rtos/zephyr/pull/65273/files#diff-00881f70fc7ec49b966fed79bd6a5a8ed5357599b241b5b029736bbc97683f83R2727), not sure of the tps service get's tested though.

A sample would be a way to showcase these LE cmds (reading local and reading remote tx power), yet the transmit power report can only be showcased by changing tx power, which I don't see any API for in zephyr (there some HCI VS commands, see [HCI_PWR_CTRL sample](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/bluetooth/hci_pwr_ctrl), which can be used with the LE Power Control Request Feature, but is not part of it.)
This is just my personal opinion, so feel free to correct me, but I don't see much worth implementing a sample that can only read tx power but doesn't act upon that info in any way.
An option to have a sample read local/remote tx power would be to include them in the HCI_PWR_CTRL sample, yet that is HCI and wouldn't use the bt_conn API's in this PR.

How to include CI integration in Zephyr is beyond my knowledge, so i'd need quite some pointers how-to. If there is a nice way to test this feature in integration, I am all for it though :)

KyraLengfeld committed:
Bluetooth: Host: Add bt shell functions LE Power Control Request Feature

To use LE commands of Le Power Control Request Feature one must utilize
the BT_TRANSMIT_POWER_CONTROL config.

Signed-off-by: Kyra Lengfeld <kyra.lengfeld@nordicsemi.no>

KyraLengfeld committed:
Bluetooth: Controller: Add Kconfig for LE Power Control Request Feature

Add the controller Kconfig to use in Zephyr.

Signed-off-by: Kyra Lengfeld <kyra.lengfeld@nordicsemi.no>

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -2905,6 +2935,26 @@ struct bt_hci_evt_le_req_peer_sca_complete {
 
+#define BT_HCI_EVT_LE_TRANSMIT_POWER_REPORT     0x21
+
+/** Reason for Transmit power reporting.
+ */
+/* Local Transmit power changed. */
+#define	BT_HCI_LE_TX_POWER_REPORT_REASON_LOCAL_CHANGED 0
+/* Remote Transmit power changed. */
+#define	BT_HCI_LE_TX_POWER_REPORT_REASON_REMOTE_CHANGED 1
+/* HCI_LE_Read_Remote_Transmit_Power_Level command completed. */
+#define	BT_HCI_LE_TX_POWER_REPORT_REASON_READ_REMOTE_COMPLETED 2
```
The HCI spec defines these in hexadecimal, so it looks a bit odd that you've decided to use decimal values. Just follow the same what the spec does:
<img width="706" alt="Screenshot 2023-11-27 at 11 57 02" src="https://github.com/zephyrproject-rtos/zephyr/assets/1032570/687eb7cf-9bf2-4710-a2bd-73920658f5d5">

jhedberg replied:
And one small thing: add an empty line between the defines and the struct

jhedberg commented:
@KyraLengfeld thanks for the updates! Mostly style & consistency related comments left - after that I'd be ready to approve. The inconsistency with using "tx" in some places and "transmit" in others bothers me a little bit as well, but not enough to block the PR (feel free to fix it though, if you want)

KyraLengfeld committed:
Bluetooth: Host: Align return lines of bt shell helper function phy2str

As other helper functions were introduced for the LE Power Control
Request Feature (see CONFIG_BT_TRANSMIT_POWER_CONTROL), it was noticed
that the return statements of switch cases should be on the next line
generally. This will align phy2str with the rest in bt.c.

Signed-off-by: Kyra Lengfeld <kyra.lengfeld@nordicsemi.no>

KyraLengfeld commented:
> @KyraLengfeld thanks for the updates! Mostly style & consistency related comments left - after that I'd be ready to approve. The inconsistency with using "tx" in some places and "transmit" in others bothers me a little bit as well, but not enough to block the PR (feel free to fix it though, if you want)

Thanks for reviewing! I amended according to your comments :) Regarding TX vs Transmit, most (if not all) names of functions/variables/defines are from how the spec calls it, and it uses both TX and Transmit (example `TX_Power_Level_Flag` parameter and `HCI_LE_Transmit_Power_Reporting` event naming). If there is anything in particular that still bothers you, let me know.

jhedberg commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -123,6 +127,67 @@ static void print_le_addr(const char *desc, const bt_addr_le_t *addr)
 #endif /* CONFIG_BT_CONN || (CONFIG_BT_BROADCASTER && CONFIG_BT_EXT_ADV) */
 
+#if defined(CONFIG_BT_TRANSMIT_POWER_CONTROL)
+static const char *txPowerFlag2str(int8_t flag)
```
Sorry, I saw this already earlier, but forgot to comment on it: no camelCase in Zephyr, please - use underscore instead (not sure why you changed the style only for some specific functions?)

KyraLengfeld replied:
I did so because I based it on the phy2str function, which isn't exactly camelCase but something in my mind thought if it had more words, it would be. I'll clean up what you commented :)

jhedberg commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -123,6 +127,67 @@ static void print_le_addr(const char *desc, const bt_addr_le_t *addr)
+	}
+}
+
+static const char *txPowerReportReason2str(uint8_t reason)
```
same here

jhedberg commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -123,6 +127,67 @@ static void print_le_addr(const char *desc, const bt_addr_le_t *addr)
+	}
+}
+
+static const char *txPwrCtrlPhy2str(enum bt_conn_le_tx_power_phy phy)
```
and here

jhedberg commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -123,6 +127,67 @@ static void print_le_addr(const char *desc, const bt_addr_le_t *addr)
+	}
+}
+
+#endif
```
Add a code comment after the `#endif` so that it's easy to see which block this terminates without having to scroll up:
```c
#endif /* CONFIG_BT_TRANSMIT_POWER_CONTROL */
```

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -2905,6 +2934,26 @@ struct bt_hci_evt_le_req_peer_sca_complete {
+	uint8_t  reason;
+	uint8_t  phy;
+	int8_t   tx_power_level;
+	int8_t   tx_power_level_flag;
```
This shouldn't be signed, so use `uint8_t`. The power level and delta can be negative, but not the flags.

jhedberg commented:
@KyraLengfeld looks generally good now, thanks! However, it still doesn't look like there's any sample or test that would actually build the new code in upstream CI. I did a quick local compilation test by forcing the new Kconfig option to be enabled, and was able to verify that the code builds without any warnings or errors. However, it would be important to have something in upstream CI doing these checks as well, so that we can be sure that the code keeps building cleanly as new changes go into Zephyr's Bluetooth stack.

KyraLengfeld commented:
> @KyraLengfeld looks generally good now, thanks! However, it still doesn't look like there's any sample or test that would actually build the new code in upstream CI. I did a quick local compilation test by forcing the new Kconfig option to be enabled, and was able to verify that the code builds without any warnings or errors. However, it would be important to have something in upstream CI doing these checks as well, so that we can be sure that the code keeps building cleanly as new changes go into Zephyr's Bluetooth stack.

Thanks! 
Would such a test be integrated into https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/bluetooth/tester ?

And would there be the option to merge it in separate PR's? (As then while I figure out the tests in rtos zephyr, I can already get going on the sdk-PRs)

jori-nordic commented:
@KyraLengfeld Just building the shell code that you added should be enough (right @jhedberg ?)

look at this: https://github.com/jori-nordic/zephyr/commit/d7c4ab605987025efb7dcd12bce8c612e435c59d

KyraLengfeld committed:
Bluetooth: Host: build shell power_control_request

(Not to mistake it for the dynamic power control nor the zephyr LL
vendor-specific commands controlling local tx power) this yaml addition
will build the shell code for the LE Power Control Feature API's, acting
as verification.

Signed-off-by: Kyra Lengfeld <kyra.lengfeld@nordicsemi.no>

jhedberg reviewed (approved):
Assuming CI passes, this is finally good to be merged, at least from my side :)
